File: app/src/main/java/org/bepass/oblivion/utils/PublicIPUtils.java
package org.bepass.oblivion.utils;

import android.os.Handler;
import android.util.Log;

import org.bepass.oblivion.model.IPDetails;
import org.json.JSONObject;

import java.net.InetSocketAddress;
import java.net.Proxy;
import java.util.Objects;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * A utility class for fetching public IP details including the country and its flag.
 */
public class PublicIPUtils {
    private static final String TAG = "PublicIPUtils";
    private static PublicIPUtils instance;
    private static final int TIMEOUT_SECONDS = 5;
    private static final int RETRY_DELAY_MILLIS = 1000;
    private static final int TIMEOUT_MILLIS = 30 * 1000;
    private final ScheduledExecutorService scheduler;
    private static final String URL_COUNTRY_API = "https://api.country.is/";

    /**
     * Private constructor to enforce singleton pattern.
     */
    private PublicIPUtils() {
        this.scheduler = Executors.newSingleThreadScheduledExecutor();
    }

    /**
     * Public method to get the singleton instance.
     *
     * @return The singleton instance of PublicIPUtils.
     */
    public static synchronized PublicIPUtils getInstance() {
        if (instance == null) {
            instance = new PublicIPUtils();
        }
        return instance;
    }

    /**
     * Fetches public IP details including the country and its flag. This method uses a
     * retry mechanism with a specified timeout to ensure reliable fetching of the IP details.
     *
     * @param callback The callback to handle the IP details once fetched.
     */
    public void getIPDetails(IPDetailsCallback callback) {
        attemptFetch(System.currentTimeMillis(), callback);
    }

    private void attemptFetch(long startTime, IPDetailsCallback callback) {
        if (System.currentTimeMillis() - startTime > TIMEOUT_MILLIS) {
            Log.d(TAG, "Timeout reached without successful IP details retrieval");
            new Handler(android.os.Looper.getMainLooper()).post(() -> callback.onDetailsReceived(new IPDetails()));
            return;
        }

        scheduler.execute(() -> {
            try {
                Log.d(TAG, "Attempting to fetch IP details");
                String portString = FileManager.getString("USERSETTING_port");
                if (portString == null || portString.isEmpty()) {
                    throw new IllegalStateException("USERSETTING_port is not set in FileManager");
                }

                int socksPort = Integer.parseInt(portString);
                Proxy proxy = new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("localhost", socksPort));

                OkHttpClient client = new OkHttpClient.Builder()
                        .proxy(proxy)
                        .connectTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
                        .readTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
                        .build();

                Request request = new Request.Builder()
                        .url(URL_COUNTRY_API)
                        .build();

                try (Response response = client.newCall(request).execute()) {
                    if (response.isSuccessful() && response.body() != null) {
                        JSONObject jsonData = new JSONObject(response.body().string());
                        IPDetails details = new IPDetails();
                        details.ip = jsonData.optString("ip");
                        details.country = jsonData.optString("country");
                        details.flag = CountryCodeUtils.toCountryFlagEmoji(new CountryCode(details.country));
                        
                        Log.d(TAG, "IP details retrieved successfully");
                        new Handler(android.os.Looper.getMainLooper()).post(() -> callback.onDetailsReceived(details));
                        return; // Success, stop here
                    }
                }
            } catch (Exception e) {
                Log.e(TAG, "Error fetching IP details: " + e.getMessage());
            }

            // If we are here, it means failure. Schedule retry.
            Log.d(TAG, "Scheduling next retry after delay");
            scheduler.schedule(() -> attemptFetch(startTime, callback), RETRY_DELAY_MILLIS, TimeUnit.MILLISECONDS);
        });
    }

    /**
     * Callback interface for receiving IP details.
     */
    public interface IPDetailsCallback {

        /**
         * Called when IP details are received.
         *
         * @param details The IP details.
         */
        void onDetailsReceived(IPDetails details);
    }
}

File: app/src/main/java/org/bepass/oblivion/utils/PublicIPUtils.kt
package org.bepass.oblivion.utils

import android.util.Log
import java.net.InetSocketAddress
import java.net.Proxy
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.bepass.oblivion.model.IPDetails
import org.json.JSONObject

/** Utility for fetching public IP details over the in-app SOCKS proxy (no network leaks). */
@Singleton
class PublicIPUtils @Inject constructor(private val okHttpClient: OkHttpClient) {
  suspend fun fetchIpDetails(): IPDetails =
    withContext(Dispatchers.IO) {
      val startTime = System.currentTimeMillis()
      var retryDelayMillis = RETRY_DELAY_MILLIS_INITIAL

      while (System.currentTimeMillis() - startTime < TIMEOUT_MILLIS) {
        try {
          return@withContext fetchOnce()
        } catch (expected: Exception) {
          Log.w(TAG, "Failed to fetch IP details; will retry", expected)
        }

        delay(retryDelayMillis)
        retryDelayMillis = (retryDelayMillis * 2).coerceAtMost(RETRY_DELAY_MILLIS_MAX)
      }

      Log.d(TAG, "Timeout reached without successful IP details retrieval")
      IPDetails()
    }

  private fun fetchOnce(): IPDetails {
    val socksPort = resolveSocksPort()
    val proxy = Proxy(Proxy.Type.SOCKS, InetSocketAddress(LOCALHOST, socksPort))

    val client =
      okHttpClient
        .newBuilder()
        .proxy(proxy)
        .connectTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
        .readTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
        .fastFallback(true) // Ensure Happy Eyeballs is enabled here too
        .build()

    val request = Request.Builder().url(URL_COUNTRY_API).build()

    client.newCall(request).execute().use { response ->
      check(response.isSuccessful) { "HTTP ${response.code} when fetching IP details" }

      val body = response.body.string()
      check(body.isNotBlank()) { "Empty response body when fetching IP details" }

      val json = JSONObject(body)
      val ip = json.optString("ip").takeIf { it.isNotBlank() }
      val country = json.optString("country").takeIf { it.isNotBlank() }
      val flag = country?.let { CountryCode(it).toCountryFlagEmoji() }?.takeIf { it.isNotBlank() }

      return IPDetails(ip = ip, country = country, flag = flag)
    }
  }

  private fun resolveSocksPort(): Int {
    val runtimePort = FileManager.getInt(FileManager.KeyHolder.RUNTIME_SOCKS_PORT, 0)
    if (runtimePort in MIN_PORT..MAX_PORT) return runtimePort

    val port =
      FileManager.getString("USERSETTING_port").toIntOrNull()?.takeIf { it in MIN_PORT..MAX_PORT }
    return port ?: error("SOCKS port is not configured (RUNTIME_SOCKS_PORT/USERSETTING_port)")
  }

  private companion object {
    private const val TAG = "PublicIPUtils"
    private const val URL_COUNTRY_API = "https://api.country.is/"
    private const val LOCALHOST = "127.0.0.1"

    private const val MIN_PORT = 1
    private const val MAX_PORT = 65535

    private const val TIMEOUT_SECONDS = 5L
    private const val TIMEOUT_MILLIS = 30_000L

    private const val RETRY_DELAY_MILLIS_INITIAL = 1_000L
    private const val RETRY_DELAY_MILLIS_MAX = 5_000L
  }
}
